+nosnoc/+discrete_time_problem/Gcs.m

```diff
@@ -0,0 +1,870 @@
+                        end
+                    end
+                  case CrossCompMode.FE_STAGE
+                    for ii=1:opts.N_stages
+                        p_stage = obj.p.p_time_var(ii);
+                        p = [p_global;p_stage];
+                        for jj=1:opts.N_finite_elements(ii);
+                            sum_lambda = lambda_prev + sum2(obj.w.lambda(ii,jj,:));
+                            Gij = {sum_lambda};
+                            Hij = {c_prev};
+                            for kk=1:(opts.n_s + rbp)
+                                x_ijk = obj.w.x(ii,jj,kk);
+                                z_ijk = obj.w.z(ii,jj,kk);
+                                c_ijk = dcs.c_fun(x_ijk,z_ijk, v_global, p);
+
+                                Gij = vertcat(Gij, {sum_lambda});
+                                Hij = vertcat(Hij, {c_ijk});
+                            end
+                            obj.G.cross_comp(ii,jj) = {vertcat(Gij{:})};
+                            obj.H.cross_comp(ii,jj) = {vertcat(Hij{:})};
+                            lambda_prev = obj.w.lambda(ii,jj,opts.n_s + rbp);
+                            x_prev = obj.w.x(ii,jj,opts.n_s + rbp);
+                            z_prev = obj.w.z(ii,jj,opts.n_s + rbp);
+                            c_prev = dcs.c_fun(x_prev,z_prev, v_global, p);
+                        end
+                    end
+                  case CrossCompMode.STAGE_FE
+                    for ii=1:opts.N_stages
+                        p_stage = obj.p.p_time_var(ii);
+                        p = [p_global;p_stage];
+                        for jj=1:opts.N_finite_elements(ii);
+                            sum_c = dcs.c_fun(x_prev);
+                            for kk=1:(opts.n_s + rbp)
+                                x_ijk = obj.w.x(ii,jj,kk);
+                                z_ijk = obj.w.z(ii,jj,kk);
+                                c_ijk = dcs.c_fun(x_ijk,z_ijk,v_global,p);
+                                sum_c = sum_c + c_ijk;
+                            end
+                            Gij = {lambda_prev};
+                            Hij = {sum_c};
+                            for kk=1:(opts.n_s + rbp)
+                                lambda_ijk = obj.w.lambda(ii,jj,kk);
+
+                                Gij = vertcat(Gij, {lambda_ijk});
+                                Hij = vertcat(Hij, {sum_c});
+                            end
+                            obj.G.cross_comp(ii,jj) = {vertcat(Gij{:})};
+                            obj.H.cross_comp(ii,jj) = {vertcat(Hij{:})};
+                            lambda_prev = obj.w.lambda(ii,jj,opts.n_s + rbp);
+                            x_prev = obj.w.x(ii,jj,opts.n_s + rbp);
+                            z_prev = obj.w.z(ii,jj,opts.n_s + rbp);
+                            c_prev = dcs.c_fun(x_prev,z_prev, v_global, p);
+                        end
+                    end
+                  case CrossCompMode.FE_FE
```

Comment:
in e.g. the msv example i get strange results with this cross comp mode, one might check is it bug free (maybe in simpler simulation problems)